src/message_data.c

```diff
@@ -0,0 +1,369 @@
+#include <nitro.h>
+#include <string.h>
+
+#include "constants/msg_data.h"
+#include "constants/narc.h"
+
+#include "inlines.h"
+
+#include "struct_decls/narc.h"
+#include "struct_decls/strbuf.h"
+
+#include "narc.h"
+#include "message_data.h"
+#include "heap.h"
+#include "strbuf.h"
+
+typedef struct {
+    u32 offset;
+    u32 len;
+} MessageParam;
+
+typedef struct MessageData_t {
+    u16 msgCount;
+    u16 rand;
+    MessageParam params[];
+} MessageData;
+
+typedef struct MessageFormatter_t {
+    u16 type;
+    u16 heapID;
+    u16 narcID;
+    u16 dataID;
+    union {
+        MessageData *data;
+        NARC *narc;
+    };
+} MessageFormatter;
+
+/**
+ * @brief Copies a source string into the destination buffer, copy as many
+ * characters as would be permitted by the input MessageParam struct.
+ * 
+ * @param dst 
+ * @param src 
+ * @param param Sources the number of characters to be copied.
+ */
+static void MessageData_Strcpy(charcode_t *dst, const charcode_t *src, const MessageParam *param);
+
+MessageData* MessageData_Load (u32 narcID, u32 dataID, u32 heapID)
+{
+    return NARC_AllocAndReadWholeMemberByIndexPair(narcID, dataID, heapID);
+}
+
+void MessageData_Free (MessageData *messageData)
+{
+    Heap_FreeToHeap(messageData);
+}
+
+static inline int CalcParamOffset (u32 msgID)
+{
+    return sizeof(MessageData) + sizeof(MessageParam) * msgID;
+}
+
+static inline void DecodeParam (MessageParam *param, u32 msgID, u32 rand)
+{
+    rand = (rand * 765 * (msgID + 1)) & 0xFFFF;
+    rand |= (rand << 16);
+
+    param->offset ^= rand;
+    param->len ^= rand;
+}
+
+static inline void DecodeString (charcode_t *str, u32 len, u32 msgID, u16 rand)
+{
+    rand = (msgID + 1) * 596947;
```

Comment:
Deserves some constants, refer to this: https://gitgud.io/gainax/pokemon_dp/-/blob/master/convert/message/msgenc.c#L5